File: java-module/src/main/java/algorithm/QuickSort.java
package algorithm;

import java.util.Collections;
import java.util.List;

public class QuickSort {

    public static void sort(List<Integer> list) {
        sort(list, 0, list.size() - 1);
    }


    private static void sort(List<Integer> list, int start, int end) {
        if (start < end) {
            int pivotIndex = partition(list, start, end);
            sort(list, start, pivotIndex - 1);
            sort(list, pivotIndex + 1, end);
        }
    }

    private static int partition(List<Integer> list, int start, int end) {
        int i = start - 1;
        int right = end;
        int pivot = list.get(end);
        for (int j = start; j < end; j++) {
            if (list.get(j) < pivot) {
                i++;
                Collections.swap(list, j, i);
            }
        }
        Collections.swap(list, i + 1, end);
        return right;
    }

}

File: scala-module/src/main/scala/algorithm/QuickSort.scala
package algorithm

object QuickSort {

  def sort(list: List[Int]): List[Int] = {
    list match {
      case pivot :: tail => {
        val lessAndMore = tail.partition(_ < pivot)
        sort(lessAndMore._1) ++ List(pivot) ++ sort(lessAndMore._2)
      }
      case Nil => Nil
    }
  }

  def main(args: Array[String]): Unit = {
    val res = sort(List(555, 1, 4, 4, 7, -1, 16))
    println(res)
  }

}
